neqo-bin/src/server/mod.rs

```diff
@@ -224,36 +225,57 @@ impl ServerRunner {
         }
     }
 
-    /// Tries to find a socket, but then just falls back to sending from the first.
-    fn find_socket(&mut self, addr: SocketAddr) -> &mut crate::udp::Socket {
-        let ((_host, first_socket), rest) = self.sockets.split_first_mut().unwrap();
-        rest.iter_mut()
-            .map(|(_host, socket)| socket)
-            .find(|socket| {
-                socket
-                    .local_addr()
-                    .ok()
-                    .map_or(false, |socket_addr| socket_addr == addr)
-            })
-            .unwrap_or(first_socket)
-    }
-
-    async fn process(&mut self, mut dgram: Option<Datagram>) -> Result<(), io::Error> {
+    async fn process(&mut self, mut ready_socket_index: Option<usize>) -> Result<(), io::Error> {
+        let mut input_dgrams: Option<DatagramIter> = None;
         loop {
-            match self.server.process(dgram.take(), (self.now)()) {
+            let input_dgram = if let Some(d) = input_dgrams.iter_mut().flatten().next() {
+                // Take datagram read in previous iteration.
+                Some(d)
+            } else {
+                // Try reading new datagrams from the socket.
+                if let Some(inx) = ready_socket_index {
+                    let (host, socket) = self.sockets.get_mut(inx).unwrap();
+                    input_dgrams = socket.recv(*host, &mut self.recv_buf)?;
+                }
+                // Then take the first datagram, if any.
+                input_dgrams.iter_mut().flatten().next().map_or_else(
+                    || {
+                        // Reading from the socket returned no datagrams. Don't try again.
+                        ready_socket_index = None;
+                        input_dgrams = None;
+                        None
+                    },
+                    Some,
+                )
+            };
+
+            // Have server process in- and output datagrams.
+            match self.server.process(input_dgram, (self.now)()) {
                 Output::Datagram(dgram) => {
-                    let socket = self.find_socket(dgram.source());
+                    let ((_host, first_socket), rest) = self.sockets.split_first_mut().unwrap();
+                    let socket = rest
+                        .iter_mut()
+                        .map(|(_host, socket)| socket)
+                        .find(|socket| {
+                            socket
+                                .local_addr()
+                                .ok()
+                                .map_or(false, |socket_addr| socket_addr == dgram.source())
+                        })
+                        .unwrap_or(first_socket);
```

Comment:
You inlined this, presumably to avoid having to pass `&mut self` to it.  It's still a useful thing to have broken out.  You can make a new function that takes `&mut self.sockets` and returns `&mut Socket`.  Either that or you could go with `Self::send(&mut self.sockets, &dgram).await?` and have that function do the sending part as well.

Reply:
You are right. Keeping this logic separate is simpler. I re-introduced `find_socket` in 786c6166.